### ISSUE TITLE
Random crashes throughout the app (RELEASE)

### ISSUE THREAD

**Issue by pavloburykh** [status-im/status-react#21032]
There are no clear steps of crash reproduction

Current crash has been reproduced on Android 12, Samsung Galaxy A52, release 2.30 (Aug 9, 2024)

Crash has happened `08-12 15:41` during building of routes while testing wallet. 

```
08-12 15:41:56.564 16587     0 E Go      : panic: runtime error: invalid memory address or nil pointer dereference
08-12 15:41:56.564 16587     0 E Go      : [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x704ae1e930]
```

[crashlogrelease.log.zip](https://github.com/user-attachments/files/16581895/crashlogrelease.log.zip)

- Status version: release 2.30 (Aug 9, 2024)
- Operating System: Android 12, Samsung Galaxy A52

**Comment by cammellos** [comment]
```
08-12 15:41:56.564 16587     0 E Go      : panic: runtime error: invalid memory address or nil pointer dereference
08-12 15:41:56.564 16587     0 E Go      : [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x704ae1e930]
08-12 15:41:56.564 16587     0 E Go      : 
08-12 15:41:56.564 16587     0 E Go      : goroutine 11852 [running]:
08-12 15:41:56.564 16587     0 E Go      : github.com/status-im/status-go/services/wallet/async.(*Scheduler).Stop.func1()
08-12 15:41:56.564 16587     0 E Go      : 	github.com/status-im/status-go/services/wallet/async/scheduler.go:197 +0x20
08-12 15:41:56.564 16587     0 E Go      : created by github.com/status-im/status-go/services/wallet/async.(*Scheduler).Stop in goroutine 11851
08-12 15:41:56.564 16587     0 E Go      : 	github.com/status-im/status-go/services/wallet/async/scheduler.go:196 +0x1b8
```

**Comment by ilmotta** [comment]
I was able to reproduce once on iOS a crash while trying to bridge. Based on the error, the code in status-go is suspicious because we are not passing the iteratee by value to the goroutine closure.

**Comment by ilmotta** [comment]
@pavloburykh, I managed to reproduce a crash with the same nil pointer error from this issue 100% of the time in an Android emulator, which is great news in terms of finding a fix.

The reproduction is:

1. Go to the bridge screen.
2. Type any value greater than zero. It doesn't matter if routes are found or not.
3. Press the back button before the calculation is done on the UI (when skeletons are still displayed).
4. App will crash with the same crash log reported in this issue.

In the video below I demonstrate it working, and then crashing.

[crash.webm](https://github.com/user-attachments/assets/5bff887e-3afe-490f-b6b1-444307bc2804)

-----

@qfrank, @friofry, please have a look at this problem to see if you can spot what could be the problem. If you could take over the issue from me even better because otherwise it will take much longer for me to figure out the solution. If you are already busy with other high-priority work please advise me on a possible solution.

I could pinpoint the line of code causing the crash https://github.com/status-im/status-go/blob/5bb7c6a8f343eb2a5e6ac93527ce31ea84ff0c2e/services/wallet/async/scheduler.go#L200, but I don't have a fix. If the line is removed, no crashes. The reason seems to be that when we delete the task from the queue there's another pending reference to the task that blows up. I also verified it's unrelated to the goroutine https://github.com/status-im/status-go/blob/5bb7c6a8f343eb2a5e6ac93527ce31ea84ff0c2e/services/wallet/async/scheduler.go#L196-L198 because you can remove it and the crash still happens, so my guess is that the reference lives outside `async.Scheduler#Stop()`.

The endpoint that initiates the crash on mobile is `wallet_stopSuggestedRoutesV2AsyncCalcualtion`, but the bug is not on the client code.

Given the problem is always reproducible on the client, it's likely possible and worth it to create a unit test in status-go to reproduce the scenario. Unfortunately, the wallet scheduler has almost no tests, so before making changes it's better to have a unit test or two, otherwise the risk of regressions can be high.

**Comment by qfrank** [comment]
> @pavloburykh, I managed to reproduce a crash with the same nil pointer error from this issue 100% of the time in an Android emulator, which is great news in terms of finding a fix.
> 
> The reproduction is:
> 
> 1. Go to the bridge screen.
> 2. Type any value greater than zero. It doesn't matter if routes are found or not.
> 3. Press the back button before the calculation is done on the UI (when skeletons are still displayed).
> 4. App will crash with the same crash log reported in this issue.
> 
> In the video below I demonstrate it working, and then crashing.
> 
>  crash.webm 
> @qfrank, @friofry, please have a look at this problem to see if you can spot what could be the problem. If you could take over the issue from me even better because otherwise it will take much longer for me to figure out the solution. If you are already busy with other high-priority work please advise me on a possible solution.
> 
> I could pinpoint the line of code causing the crash https://github.com/status-im/status-go/blob/5bb7c6a8f343eb2a5e6ac93527ce31ea84ff0c2e/services/wallet/async/scheduler.go#L200, but I don't have a fix. If the line is removed, no crashes. The reason seems to be that when we delete the task from the queue there's another pending reference to the task that blows up. I also verified it's unrelated to the goroutine https://github.com/status-im/status-go/blob/5bb7c6a8f343eb2a5e6ac93527ce31ea84ff0c2e/services/wallet/async/scheduler.go#L196-L198 because you can remove it and the crash still happens, so my guess is that the reference lives outside `async.Scheduler#Stop()`.
> 
> The endpoint that initiates the crash on mobile is `wallet_stopSuggestedRoutesV2AsyncCalcualtion`, but the bug is not on the client code.
> 
> Given the problem is always reproducible on the client, it's likely possible and worth it to create a unit test in status-go to reproduce the scenario. Unfortunately, the wallet scheduler has almost no tests, so before making changes it's better to have a unit test or two, otherwise the risk of regressions can be high.

leave it to me, I can have a look 👀 @ilmotta

**Comment by pavloburykh** [comment]
> @pavloburykh, I managed to reproduce a crash with the same nil pointer error from this issue 100% of the time in an Android emulator, which is great news in terms of finding a fix.

Wow, great job @ilmotta! Thank you for finding the steps. They worked for me as well.

**Comment by VolodLytvynenko** [comment]
Hi @ilmotta @qfrank I've encountered one more crash that always happens when I try to send a collectible. Could you verify if this crash is the same as a reported one or if it's caused by different factors? Just want to avoid creating duplicates in GH repo

https://github.com/user-attachments/assets/3dafc89b-ec15-46d3-8baa-7953e367d329

[crash.txt](https://github.com/user-attachments/files/16625701/crash.txt)

**Comment by pavloburykh** [comment]
> @pavloburykh, I managed to reproduce a crash with the same nil pointer error from this issue 100% of the time in an Android emulator, which is great news in terms of finding a fix.

@ilmotta just to make sure: I see this fix https://github.com/status-im/status-go/pull/5710 has been merged before you identified the steps and got some additional details on crash. Is this fix still relevant? As far as I understand it is not fixing crash but it still fixing some other problem, correct?

**Comment by qfrank** [comment]
> Hi @ilmotta @qfrank I've encountered one more crash that always happens when I try to send a collectible. Could you verify if this crash is the same as a reported one or if it's caused by different factors? Just want to avoid creating duplicates in GH repo
> 
>  collect.mp4 
> [crash.txt](https://github.com/user-attachments/files/16625701/crash.txt)

```
16:17:12.629  E  panic: runtime error: invalid memory address or nil pointer dereference
16:17:12.629  E  github.com/status-im/status-go/services/wallet/router.(*Router).getBalanceMapForTokenOnChains(0x4001761180, {0x77d938a988, 0x4003c52f50}, 0x4000382ff0, {0x4005ba8e60, 0x3, 0x4003dfa278?})
16:17:12.629  E          github.com/status-im/status-go/services/wallet/router/router_v2.go:557 +0x53c
16:17:12.629  E  github.com/status-im/status-go/services/wallet/router.(*Router).SuggestedRoutesV2(0x4001761180, {0x77d938a988, 0x4003c52f50}, 0x4000382ff0)
16:17:12.629  E          github.com/status-im/status-go/services/wallet/router/router_v2.go:501 +0x134
16:17:12.629  E  github.com/status-im/status-go/services/wallet/router.(*Router).SuggestedRoutesV2Async.func1({0x77d938a988?, 0x4003c52f50?})
16:17:12.629  E          github.com/status-im/status-go/services/wallet/router/router_v2.go:450 +0x38
16:17:12.629  E  github.com/status-im/status-go/services/wallet/async.(*Scheduler).runTask.func1()
16:17:12.629  E          github.com/status-im/status-go/services/wallet/async/scheduler.go:141 +0x48
16:17:12.629  E  created by github.com/status-im/status-go/services/wallet/async.(*Scheduler).runTask in goroutine 5120
16:17:12.629  E          github.com/status-im/status-go/services/wallet/async/scheduler.go:140 +0x130
```

the invocation stack is different from [here mentioned](https://github.com/status-im/status-mobile/issues/21032#issuecomment-2283928497) before , so it should be different @VolodLytvynenko

**Comment by VolodLytvynenko** [comment]
> > Hi @ilmotta @qfrank I've encountered one more crash that always happens when I try to send a collectible. Could you verify if this crash is the same as a reported one or if it's caused by different factors? Just want to avoid creating duplicates in GH repo
> > collect.mp4
> > [crash.txt](https://github.com/user-attachments/files/16625701/crash.txt)
> 
> ```
> 16:17:12.629  E  panic: runtime error: invalid memory address or nil pointer dereference
> 16:17:12.629  E  github.com/status-im/status-go/services/wallet/router.(*Router).getBalanceMapForTokenOnChains(0x4001761180, {0x77d938a988, 0x4003c52f50}, 0x4000382ff0, {0x4005ba8e60, 0x3, 0x4003dfa278?})
> 16:17:12.629  E          github.com/status-im/status-go/services/wallet/router/router_v2.go:557 +0x53c
> 16:17:12.629  E  github.com/status-im/status-go/services/wallet/router.(*Router).SuggestedRoutesV2(0x4001761180, {0x77d938a988, 0x4003c52f50}, 0x4000382ff0)
> 16:17:12.629  E          github.com/status-im/status-go/services/wallet/router/router_v2.go:501 +0x134
> 16:17:12.629  E  github.com/status-im/status-go/services/wallet/router.(*Router).SuggestedRoutesV2Async.func1({0x77d938a988?, 0x4003c52f50?})
> 16:17:12.629  E          github.com/status-im/status-go/services/wallet/router/router_v2.go:450 +0x38
> 16:17:12.629  E  github.com/status-im/status-go/services/wallet/async.(*Scheduler).runTask.func1()
> 16:17:12.629  E          github.com/status-im/status-go/services/wallet/async/scheduler.go:141 +0x48
> 16:17:12.629  E  created by github.com/status-im/status-go/services/wallet/async.(*Scheduler).runTask in goroutine 5120
> 16:17:12.629  E          github.com/status-im/status-go/services/wallet/async/scheduler.go:140 +0x130
> ```
> 
> the invocation stack is different from [here mentioned](https://github.com/status-im/status-mobile/issues/21032#issuecomment-2283928497) before , so it should be different @VolodLytvynenko

@qfrank thanx for clarification

**Comment by ilmotta** [comment]
> > @pavloburykh, I managed to reproduce a crash with the same nil pointer error from this issue 100% of the time in an Android emulator, which is great news in terms of finding a fix.
> 
> @ilmotta just to make sure: I see this fix [status-im/status-go#5710](https://github.com/status-im/status-go/pull/5710) has been merged before you identified the steps and got some additional details on crash. Is this fix still relevant? As far as I understand it is not fixing crash but it still fixing some other problem, correct?

Good point, forgot to mention this. That fix is not necessary in the release branch, but if we add it's harmless @pavloburykh
